frame/composable-traits/src/undercollateralized_loans/configs.rs

```diff
@@ -0,0 +1,299 @@
+use crate::defi::CurrencyPair;
+use frame_support::pallet_prelude::*;
+use sp_runtime::Perquintill;
+use sp_std::{
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	vec::Vec,
+};
+
+#[derive(Encode, Decode, Default, TypeInfo, RuntimeDebug, Clone, Eq, PartialEq)]
+pub struct MarketConfig<AccountId, AssetId, BlockNumber, VaultId>
+where
+	AccountId: Clone + Eq + PartialEq,
+	AssetId: Clone + Eq + PartialEq,
+	BlockNumber: Clone + Eq + PartialEq,
+	VaultId: Clone + Eq + PartialEq,
+{
+	/// The account id of this market
+	account_id: AccountId,
+	/// The owner of this market.
+	manager: AccountId,
+	/// The vault containing the borrow asset.
+	borrow_asset_vault: VaultId,
+	/// The asset being used as collateral.
+	collateral_asset_id: AssetId,
+	/// The asset being used as borrow asset.
+	borrow_asset_id: AssetId,
+	/// Number of blocks until invalidate oracle's price.
+	max_price_age: BlockNumber,
+	/// Borrowers which are allowed to use the service.
+	whitelist: BTreeSet<AccountId>,
+}
+
+impl<AccountId, AssetId, BlockNumber, VaultId>
+	MarketConfig<AccountId, AssetId, BlockNumber, VaultId>
+where
+	AccountId: Clone + Eq + PartialEq,
+	AssetId: Clone + Eq + PartialEq,
+	BlockNumber: Clone + Eq + PartialEq,
+	VaultId: Clone + Eq + PartialEq,
+{
+	pub fn new(
+		account_id: AccountId,
+		manager: AccountId,
+		borrow_asset_vault: VaultId,
+		borrow_asset_id: AssetId,
+		collateral_asset_id: AssetId,
+		max_price_age: BlockNumber,
+		whitelist: BTreeSet<AccountId>,
+	) -> Self {
+		Self {
+			account_id,
+			manager,
+			borrow_asset_vault,
+			borrow_asset_id,
+			collateral_asset_id,
+			max_price_age,
+			whitelist,
+		}
+	}
+
+	/// Get a reference to the market config's account id.
+	pub fn account_id(&self) -> &AccountId {
+		&self.account_id
+	}
+
+	/// Get a reference to the market config's manager.
+	pub fn manager(&self) -> &AccountId {
+		&self.manager
+	}
+	/// Get a reference to the market config's borrow asset vault.
+	pub fn borrow_asset_vault(&self) -> &VaultId {
+		&self.borrow_asset_vault
+	}
+
+	/// Get a reference to the market config's borrow asset.
+	pub fn borrow_asset_id(&self) -> &AssetId {
+		&self.borrow_asset_id
+	}
+
+	/// Get a reference to the market config's collateral asset.
+	pub fn collateral_asset_id(&self) -> &AssetId {
+		&self.collateral_asset_id
+	}
+
+	/// Get a reference to the market config's max price age.
+	pub fn max_price_age(&self) -> &BlockNumber {
+		&self.max_price_age
+	}
+
+	/// Get a reference to the market config's whitelist.
+	#[must_use]
+	pub fn whitelist(&self) -> &BTreeSet<AccountId> {
+		&self.whitelist
+	}
+}
+
+#[derive(Encode, Decode, Default, TypeInfo, RuntimeDebug, Clone, Eq, PartialEq)]
+pub struct LoanConfig<AccountId, AssetId, Balance, Timestamp>
```

Comment:
I'm wondering if it's necessary to have private fields with getters? What imvariants are being upheld?

Reply:
(For all types in this file)

Reply:
This is just to emphasise that these fields must be immutable.

Reply:
I see, I think it would be good to document that these structs are readonly and why they are

Reply:
fixed